### PR TITLE
fix(store): refresh gRPC channels after address changes

### PR DESCRIPTION
**Pull request by bitflicker64**
## Purpose of the PR

> Kept in sync with https://github.com/apache/hugegraph/pull/3130

HStore caches gRPC channels and blocking/asynchronous stubs by Store target.
When a stable target resolves to a different address, those pools can retain the
failed transport indefinitely even though the replacement Store is reachable.

This is the channel-lifecycle half of #3124. Complete recovery behind a stable
DNS name still requires the finite positive DNS TTL from #3126; this patch does
not bypass an indefinitely stale JVM DNS cache.

## Main Changes

- Fingerprint each target's sorted, deduplicated resolved address set before
  selecting a channel, with explicit first-run state and overflow-safe refresh
  deadlines.
- Keep DNS resolution and cold channel initialization off restricted Gremlin
  callers using bounded prestarted executors (4 maintenance, 4 initialization,
  and 1 retirement thread), with independent targets initialized concurrently.
- Publish complete replacement pools before retiring old channels; preserve the
  healthy pool across resolution, submission, creation, cleanup, and fatal-error
  paths, and force-terminate every partial pool before propagating failure.
- Rebuild blocking and asynchronous stub pools only from the currently published
  channel pool; route QueryV2 through that guarded path and clean up injected test
  channels when they are replaced.
- Add client/target lifecycle cleanup for channels, stubs, refresh metadata, and
  locks. Terminal Store notices evict exact node identities, stale sessions cannot
  resurrect state, and late notices cannot evict a same-address replacement.
- Parse raw `host:port`, `dns:///host:port`, and bracketed IPv6 targets for
  fingerprinting; unsupported resolver schemes remain delegated to gRPC.
- Add focused concurrency and failure regressions covering SecurityManager
  boundaries, target isolation, lifecycle races, partial creation, interrupts,
  executor shutdown, Store-node eviction, QueryV2, and DNS deadline edge cases.

## Verifying these changes

- [x] Need tests and can be verified as follows:
  - Java 11: `mvn test -pl hugegraph-store/hg-store-test -am -Dtest=AbstractGrpcClientTest -Dsurefire.failIfNoSpecifiedTests=false`: 37 tests, 0 failures, 0 errors, 0 skips.
  - `git diff --check fad20a3c515018278869cc62e8162894373bff95..bdd8df11106d8b015228655ed140e70543469738`: passed with no output.
  - Three independent exact-head reviews and one Design Audit completed with no blocking findings.

## Does this PR potentially affect the following parts?

- [ ] Dependencies
- [ ] Modify configurations
- [x] The public API (adds explicit channel/client lifecycle operations)
- [ ] Other affects
- [ ] Nope

## Documentation Status

- [ ] `Doc - TODO`
- [ ] `Doc - Done`
- [x] `Doc - No Need`

## Blocking review follow-up (2026-08-13)

- Exact fix commit: `bdd8df11106d8b015228655ed140e70543469738`, directly on prior PR head `fad20a3c515018278869cc62e8162894373bff95`.
- Closed the three latest blocking findings: eager thread footprint, permanent target-state retention, and partial/null publication after `Error`.
- Also hardened close/register/refresh races, stale-node replacement identity, executor-shutdown convergence, and interrupt restoration.

## Summary by CodeRabbit

- **功能改进**
  - 优化连接与会话管理，节点或连接发生变化后可自动重新绑定，提升请求连续性。
  - 增强节点不可用、连接关闭及 RPC 取消场景的处理，减少错误连接和失效会话。
  - 支持按目标关闭连接资源，便于连接释放与故障恢复。
  - 改进并发场景下的会话复用，避免重复创建并确保使用最新节点。
- **稳定性**
  - 增强节点替换和过期通知处理，降低请求发送至旧节点的风险。

### CODE REVIEW

File: hugegraph-store/hg-store-client/src/main/java/org/apache/hugegraph/store/client/grpc/AbstractGrpcClient.java [hunk 1]
```diff
@@ -169,6 +268,323 @@ protected AbstractStub setStubOption(AbstractStub value) {
...
+     * followed by a path, so {@code unix:/var/run/store.sock} is a scheme while
+     * {@code store:8500} is not.
+     */
+    private static boolean hasResolverScheme(String target) {
+        int scheme = target.indexOf(':');
+        return scheme >= 0 && scheme + 1 < target.length() && target.charAt(scheme + 1) == '/';
+    }
+
+    protected InetAddress[] resolveHost(String host) throws UnknownHostException {
+        return InetAddress.getAllByName(host);
```
[the remark this answers is not in the document]
Comment (bitflicker64) [reply hugegraph-store/hg-store-client/src/main/java/org/apache/hugegraph/store/client/grpc/AbstractGrpcClient.java]: Valid limitation and explicitly documented in the PR scope: this channel-lifecycle change polls InetAddress.getAllByName(), but it cannot bypass an indefinitely cached positive JVM DNS result. The PR description states that complete recovery behind a stable DNS name requires a finite positive DNS TTL from #3126. This should be addressed by the launcher/configuration or the resolver work in that issue; changing channel refresh alone would not make the JVM return a new address.

File: hugegraph-store/hg-store-client/src/main/java/org/apache/hugegraph/store/client/grpc/AbstractGrpcClient.java [hunk 2]
```diff
@@ -169,6 +232,316 @@ protected AbstractStub setStubOption(AbstractStub value) {
...
+    }
+
+    void submitChannelCreation(Runnable task) {
+        this.executor.execute(task);
+    }
+
+    private void retireChannels(ManagedChannel[] retiredChannels) {
+        Arrays.stream(retiredChannels)
+              .filter(channel -> channel != null && !channel.isShutdown())
+              .forEach(ManagedChannel::shutdown);
```
[the remark this answers is not in the document]
Comment (bitflicker64) [reply hugegraph-store/hg-store-client/src/main/java/org/apache/hugegraph/store/client/grpc/AbstractGrpcClient.java]: Valid concern. The old pool is published for replacement and then gracefully shutdown immediately; a stub handed out just before that publication can still have its first RPC rejected by gRPC shutdown. The current tests cover active-call draining and the publication/retirement race, but not a post-return pre-first-call grace lease. This is a real follow-up for retry/lease semantics rather than a DNS refresh correctness fix, so it remains outside this PR scope.

File: hugegraph-store/hg-store-client/src/main/java/org/apache/hugegraph/store/client/grpc/AbstractGrpcClient.java [hunk 3]
```diff
@@ -58,31 +119,52 @@ public AbstractGrpcClient() {
...
+                    if (interruption != null) {
+                        forceTerminateChannels(tc);
+                        Thread.currentThread().interrupt();
+                        throw new RuntimeException(interruption);
+                    }
+                } catch (Throwable e) {
+                    throw e instanceof RuntimeException ? (RuntimeException) e :
+                          new RuntimeException(e);
                 }
+                channels.put(target, tc);
```
[the remark this answers is not in the document]
Comment (bitflicker64) [reply hugegraph-store/hg-store-client/src/main/java/org/apache/hugegraph/store/client/grpc/AbstractGrpcClient.java]: Valid scalability observation. The cold path currently waits for initialization while holding the global channels monitor, so a slow first target can serialize other cold starts. The dedicated initialization executor prevents DNS-refresh starvation and preserves the restricted-caller guarantee, but it does not remove this monitor scope. Removing the wait from the global monitor is a separate concurrency refactor; no production behavior change is being claimed for that in this PR.

File: hugegraph-store/hg-store-client/src/main/java/org/apache/hugegraph/store/client/grpc/AbstractGrpcClient.java [hunk 4]
```diff
@@ -169,6 +268,323 @@ protected AbstractStub setStubOption(AbstractStub value) {
...
+        if (target == null || target.isEmpty()) {
+            return "";
+        }
+
+        String endpoint = target;
+        if (target.startsWith(DNS_SCHEME)) {
+            endpoint = target.substring(DNS_SCHEME.length());
+            if (!endpoint.startsWith("/")) {
+                return "";
+            }
```
[the remark this answers is not in the document]
Comment (bitflicker64) [reply hugegraph-store/hg-store-client/src/main/java/org/apache/hugegraph/store/client/grpc/AbstractGrpcClient.java]: Agreed. The implementation and test intentionally reject the opaque dns:host:port form and uppercase DNS, while accepting raw host:port, dns:///host:port, and dns://authority/host:port. The Javadoc still describes the rejected opaque form and should be corrected to match the current parser; this is documentation-only and does not indicate a parser behavior bug.
